Sources/X509/Verifier/VerificationDiagnostic.swift

```diff
@@ -0,0 +1,237 @@
+        ))
+    }
+}
+
+extension VerificationDiagnostic.Storage {
+    static func leafCertificateHasUnhandledCriticalExtension(
+        _ leafCertificate: Certificate,
+        handledCriticalExtensions: [ASN1ObjectIdentifier]
+    ) -> Self {
+        .leafCertificateHasUnhandledCriticalExtension(.init(
+            leafCertificate: leafCertificate,
+            handledCriticalExtensions: handledCriticalExtensions
+        ))
+    }
+    static func leafCertificateIsInTheRootStoreButDoesNotMeetPolicy(
+        _ leafCertificate: Certificate,
+        reason failsToMeetPolicyReason: String
+    ) -> Self {
+        .leafCertificateIsInTheRootStoreButDoesNotMeetPolicy(.init(
+            leafCertificate: leafCertificate,
+            failsToMeetPolicyReason: failsToMeetPolicyReason
+        ))
+    }
+    static func chainFailsToMeetPolicy(
+        _ chain: UnverifiedCertificateChain,
+        reason failsToMeetPolicyReason: String
+    ) -> Self {
+        .chainFailsToMeetPolicy(.init(
+            chain: chain,
+            failsToMeetPolicyReason: failsToMeetPolicyReason
+        ))
+    }
+    static func issuerHashUnhandledCriticalExtension(
+        issuer: Certificate,
+        partialChain: [Certificate],
+        handledCriticalExtensions: [ASN1ObjectIdentifier]
+    ) -> Self {
+        .intermediateHashUnhandledCriticalExtension(.init(
+            issuer: issuer,
+            partialChain: partialChain,
+            handledCriticalExtensions: handledCriticalExtensions
+        ))
+    }
+    static func issuerHasNotSignedCertificate(
+        _ issuer: Certificate,
+        partialChain: [Certificate]
+    ) -> Self {
+        .issuerHasNotSignedCertificate(.init(
+            issuer: issuer,
+            partialChain: partialChain
+        ))
+    }
+}
+
+extension VerificationDiagnostic: CustomStringConvertible {
+    /// Produces a human readable description of this ``VerificationDiagnostic`` that is potential expensive to compute.
+    public var description: String {
+        storage.description
+    }
+}
+
+extension VerificationDiagnostic.Storage {
+    var description: String {
+        switch self {
+        case .leafCertificateHasUnhandledCriticalExtension(let diagnostic): return diagnostic.description
+        case .leafCertificateIsInTheRootStoreButDoesNotMeetPolicy(let diagnostic): return diagnostic.description
+        case .chainFailsToMeetPolicy(let diagnostic): return diagnostic.description
+        case .intermediateHashUnhandledCriticalExtension(let diagnostic): return diagnostic.description
+        case .issuerHasNotSignedCertificate(let diagnostic): return diagnostic.description
+        }
+    }
+}
+
+extension VerificationDiagnostic.LeafCertificateHasUnhandledCriticalExtensions: CustomStringConvertible {
+    var description: String {
+        """
+        The leaf certificate has critical extensions that the policy does not understand and therefore can't enforce.
```

Comment:
Given this is multiline output, it might make sense to add new lines in appropriate places to make it more readable:

```swift
"""
The leaf certificate ... enforce.

Unhandled extensions:
...

Leaf certificate:
...
"""
```

Further to that I don't know what the format of the unhandled extensions is but might make sense to bullet point each.

Reply:
Yep, lets do that.